logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,158 @@
+    }
+
+    private fun enqueueMessageDeletion(message: Message.Ephemeral) {
+        userSessionCoroutineScope.launch {
+            val selfDeletingMessageTimer = SelfDeletingMessageTimer(
+                coroutineScope = userSessionCoroutineScope
+            )
+
+            for (timerEvent in selfDeletingMessageTimer.startTimer(message.timeLeftForDeletion())) {
+                when (timerEvent) {
+                    is SelfDeletionTimerState.Started -> {
+                        if (!message.isDeletionScheduledInThePast()) {
+                            messageRepository.markSelfDeletionStartDate(
+                                message.conversationId,
+                                message.id,
+                                Clock.System.now().toEpochMilliseconds()
+                            )
+                        }
+                    }
+
+                    is SelfDeletionTimerState.OnGoing -> {
+                        println("time left ${timerEvent.timeLeft.timeLeftInMs}")
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState + ((message.conversationId to message.id) to timerEvent.timeLeft)
+                            }
+                        }
+                    }
+
+                    is SelfDeletionTimerState.Finished -> {
+                        messageRepository.deleteMessage(message.id, message.conversationId)
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState - (message.conversationId to message.id)
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+    }
+
+    override fun observePendingMessageDeletionState() = ephemeralMessageDeletionTimeLeftMap
+
+    override fun enqueuePendingSelfDeletionMessages() {
+        launch {
+            messageRepository.getEphemeralMessages()
+                .onSuccess { ephemeralMessages ->
+                    ephemeralMessages.forEach { ephemeralMessage ->
+                        require(ephemeralMessage is Message.Ephemeral)
+
+                        enqueueMessageDeletion(message = ephemeralMessage)
+                    }
+                }
+        }
+    }
+
+}
+
+internal class SelfDeletingMessageTimer(
+    private val coroutineScope: CoroutineScope
+) : CoroutineScope by coroutineScope {
+    private companion object {
+        const val TIMER_UPDATE_INTERVAL_IN_MILLIS = 1000L
+    }
+
+    fun startTimer(expireAfterMillis: Long) = produce {
+        send(SelfDeletionTimerState.Started(Clock.System.now().toEpochMilliseconds()))
+
+        var elapsedTime = 0L
+
+        while (elapsedTime < expireAfterMillis) {
+            delay(TIMER_UPDATE_INTERVAL_IN_MILLIS)
```

Comment:
Why do we need to update every second? Why not delay the remaining time until message should be deleted? 

Is the intention that the UI should also observe these timers? Because I think UI should manager their own timers.

I think this can simplified and we can get rid of the mutex locks.

Reply:
Also mutexes will still be needed, because someone can request the deletion of the message couple of times for the same message, we would need to hold the map somewhere with the ones that are ongoing to prevent that, and we will also need a mutex or some concurency protection mechanism when multiple coroutines are trying to create a timer, it will be I think one less place of locking the access with mutex

Reply:
But what does the UI do with these timers? It displays a countdown of some sort? Such a timer would only need to run for a message which actually visible on screen.  What if the UI wants to update the count down every millisecond?

Managing such display timer in Kalium means we'll need to run timers for all ephemeral messages regardless if they displayed or not.

Reply:
I mean if that would be a real requirement we could add a variable that defines how often we want to see the update, and yes it display as countdown, the process of self-deletion starts when the message is displayed on the screen, there is a use case that allows to do so, and yes we indeed run a timer if the messages is displayed or not but I do not think it is a huge problem to be honest, it simply updates a value that is held inside a map

Reply:
I cannot see the benefit of doing the work to move this all to the UI and add another mechanism in the UI that duplicates basicly what is done here, just for the sake of not updating a value when it is not being displayed

Reply:
However now you are saying it would not make much sense to have a timer running each second for a message deleted after 2 weeks

Reply:
I am going to take another look on this with what u are saying in mind I think you got a point here

Reply:
> I mean if that would be a real requirement we could add a variable that defines how often we want to see the update, and yes it display as countdown, the process of self-deletion starts when the message is displayed on the screen, there is a use case that allows to do so, and yes we indeed run a timer if the messages is displayed or not but I do not think it is a huge problem to be honest, it simply updates a value that is held inside a map

In Android you want to display a countdown, maybe another client wants to display a continuous progress bar. I agree it might seem like are implementing something twice but I'd argue visual timers and a timer for deleting a message have different requirements and we shouldn't try to merge them.

Reply:
What they want to display does not mean we should remove it from the Kalium I think, they are interested in progress and the time that is left, forcing them to implement something on the client side if we have all the information on Kalium is unncessary I think if we can have this built in, I am going to find a solution so we are both happy :P